pkg/imgpkg/bundle/images_refs.go

```diff
@@ -0,0 +1,188 @@
+
+type ImageRef struct {
+	lockconfig.ImageRef
+	IsBundle *bool
+}
+
+func NewImageRef(imgRef lockconfig.ImageRef, isBundle bool) ImageRef {
+	return ImageRef{ImageRef: imgRef, IsBundle: &isBundle}
+}
+
+func (i ImageRef) DeepCopy() ImageRef {
+	return ImageRef{
+		ImageRef: i.ImageRef.DeepCopy(),
+		IsBundle: i.IsBundle,
+	}
+}
+func (i *ImageRef) updateRepo(relativeToRepo string) *ImageRef {
+	i.Image = i.replaceImageRepo(relativeToRepo)
+	return i
+}
+func (i *ImageRef) replaceImageRepo(relativeToRepo string) string {
+	imgParts := strings.Split(i.Image, "@")
+	if len(imgParts) != 2 {
+		panic(fmt.Sprintf("Internal inconsistency: The provided image URL '%s' does not contain a digest", i.Image))
+	}
+	updatedRepo := relativeToRepo + "@" + imgParts[1]
+	return updatedRepo
+}
+
+type ImageRefs struct {
+	refs                 []ImageRef
+	imageLocationsConfig *ImageLocationsConfig
+	originalImagesLock   lockconfig.ImagesLock
+
+	*sync.Mutex
+}
+
+func NewEmptyImageRefs() ImageRefs {
+	return ImageRefs{
+		Mutex: &sync.Mutex{},
+	}
+}
+
+func NewImageRefs(imagesLock lockconfig.ImagesLock, imagesLockLocationConfig ImagesLockLocationsConfig) (ImageRefs, error) {
+	imageRefs := ImageRefs{
+		Mutex:              &sync.Mutex{},
+		originalImagesLock: imagesLock,
+	}
+	for _, lockImgRef := range imagesLock.Images {
+		imageRefs.AddImagesRef(ImageRef{
+			ImageRef: lockImgRef,
+			IsBundle: nil,
+		})
+	}
+
+	err := imageRefs.syncImageRefsWithLocationConfig(imagesLockLocationConfig)
+	if err != nil {
+		return ImageRefs{}, err
+	}
+	return imageRefs, nil
+}
+
+func (i *ImageRefs) LocalizeToBundle(relativeToRepo string) {
+	for _, imgRef := range i.refs {
+		imgRef.AddLocation(imgRef.replaceImageRepo(relativeToRepo))
+		i.AddImagesRef(imgRef)
+	}
+}
+
+func (i *ImageRefs) UpdateRelativeToRepo(imgRetriever ctlimg.ImagesMetadata, relativeToRepo string) (bool, error) {
+	if i.imageLocationsConfig != nil {
+		i.LocalizeToBundle(relativeToRepo)
+		return true, nil
+	}
+
+	for _, ref := range i.refs {
+		image, err := name.ParseReference(ref.updateRepo(relativeToRepo).Image)
+		if err != nil {
+			return false, err
+		}
+		_, err = imgRetriever.Digest(image)
+		if err != nil {
+			if terr, ok := err.(*transport.Error); ok {
+				if _, ok := imageNotFoundStatusCode[terr.StatusCode]; ok {
+					return false, nil
+				}
+			}
+			return false, err
+		}
+	}
+
+	i.LocalizeToBundle(relativeToRepo)
+
+	return true, nil
+}
+func (i *ImageRefs) AddImagesRef(refs ...ImageRef) {
+	i.Lock()
+	defer i.Unlock()
+
+	for _, ref := range refs {
+		found := false
+		for j, imageRef := range i.refs {
+			if imageRef.Image == ref.Image {
+				found = true
+				i.refs[j] = ref.DeepCopy()
+				break
+			}
+		}
+
+		if !found {
+			i.refs = append(i.refs, ref)
+		}
+	}
+}
+func (i *ImageRefs) Find(ref string) (ImageRef, bool) {
+	for _, imageRef := range i.refs {
+		if imageRef.Image == ref {
+			return imageRef, true
```

Comment:
I don't see why.

Reply:
in general its typically good to return copies of objects to avoid accidental mutation down the line. in this case each ImageRef contains some mutable data which we wouldnt want any caller to modify (e.g. AddLocations on lockconfig.ImageRef for example) and affect internal state of ImageRefs struct.

Reply:
i'll make the change.